### ISSUE TITLE
Connection fail to work after fork

### ISSUE THREAD

**Issue by fiscal-cliff** [jeremyevans/sequel#1431]
### Complete Description of Issue
There is a bug in sequel connection pool. It does not track PID changes, like Active Record does.
https://stackoverflow.com/questions/29918475/unicorn-do-i-need-to-disconnect-from-database-in-before-fork-block
If you didn't disconnect the connection in before_fork callback you would get errors related to database adapter. And it took me some time to realise that it wasn't coincident. I have found the issue [#691](https://github.com/jeremyevans/sequel/issues/691), and I quess the detection of PID changes could be useful.

Thank you for sequel!

### Simplest Possible Self-Contained Example Showing the Bug

5.times { DB[:event_log_entries].count; Process.fork { DB[:event_log_entries].count } }

### Full Backtrace of Exception (if any)

A PG::ConnectionBad occurred in background 

 **PQsocket() can't get socket descriptor**
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/adapters/base.rb:63:in `async_exec'

-------------------------------
Backtrace:
-------------------------------

bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/adapters/base.rb:63:in `async_exec'
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/adapters/base.rb:63:in `block in execute_sql'
bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/database/connecting.rb:301:in `block in synchronize'
bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/connection_pool/threaded.rb:103:in `hold'
bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/database/connecting.rb:301:in `synchronize'
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/adapters/sequel.rb:12:in `checkout'
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/adapters/base.rb:63:in `execute_sql'
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/adapters/base.rb:70:in `block in execute_prepared'
bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/database/connecting.rb:301:in `block in synchronize'
bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/connection_pool/threaded.rb:103:in `hold'
bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/database/connecting.rb:301:in `synchronize'
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/adapters/sequel.rb:12:in `checkout'
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/adapters/base.rb:67:in `execute_prepared'
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/adapters/base.rb:50:in `execute'
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que.rb:68:in `execute'
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/job.rb:113:in `block in work'
bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/database/connecting.rb:301:in `block in synchronize'
bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/connection_pool/threaded.rb:107:in `hold'
bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/database/connecting.rb:301:in `synchronize'
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/adapters/sequel.rb:12:in `checkout'
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/job.rb:111:in `work'
bundle/ruby/2.4.0/bundler/gems/ptero-share-eb32558250d7/lib/ptero_share/extensions/que.rb:70:in `block in work'
bundle/ruby/2.4.0/gems/lamian-1.0.1/lib/lamian.rb:47:in `block in run'
bundle/ruby/2.4.0/gems/lamian-1.0.1/lib/lamian/logger.rb:33:in `run'
bundle/ruby/2.4.0/gems/lamian-1.0.1/lib/lamian.rb:47:in `run'
bundle/ruby/2.4.0/bundler/gems/ptero-share-eb32558250d7/lib/ptero_share/extensions/que.rb:70:in `work'
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/worker.rb:80:in `block in work_loop'
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/worker.rb:75:in `loop'
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/worker.rb:75:in `work_loop'
bundle/ruby/2.4.0/gems/que-0.14.0/lib/que/worker.rb:19:in `block in initialize'

**Comment by jeremyevans** [comment]
This isn't a bug, it's a specific design decision.  You need to disconnect before fork, not try to ignore the problem after fork.  See http://sequel.jeremyevans.net/rdoc/files/doc/code_order_rdoc.html#label-Disconnect+If+Using+Forking+Webserver+with+Code+Preloading

You should consider that forking at all with database connections breaks the connections in the parent process:

```ruby
DB = Sequel.connect('postgres:///')
DB.tables
Process.wait(fork{})
DB.tables
```

The same behavior occurs with ActiveRecord:

```ruby
ActiveRecord::Base.establish_connection(:adapter=>'postgresql')
ActiveRecord::Base.connection.execute('SELECT 1')
Process.wait(fork{})
ActiveRecord::Base.connection.execute('SELECT 1')
```

In both cases, you end up with an wrapped error for:

```
PG::ConnectionBad: PQconsumeInput() server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
```

Basically, ActiveRecord tries to work around buggy user code in a particular use case, while Sequel does not.

**Comment by eregon** [comment]
FWIW, it seems ActiveRecord 5.2.0 succeeds that test now:
```ruby
require 'active_record'
ActiveRecord::Base.establish_connection(adapter: 'postgresql', database: 'test')
ActiveRecord::Base.connection.execute('SELECT 1')
Process.wait(fork{
  puts "In child"
})
ActiveRecord::Base.connection.execute('SELECT 1')
puts "Done"
```
gives
```
In child
Done
```

While Sequel fails it:
```ruby
require 'sequel'
DB = Sequel.connect('postgres:///test')
DB.tables
Process.wait(fork{
  puts "In child"
})
DB.tables
puts "Done"
```
gives
```
In child
gems/sequel-5.10.0/lib/sequel/adapters/postgres.rb:152:in `async_exec': PG::ConnectionBad: PQconsumeInput() server closed the connection unexpectedly (Sequel::DatabaseDisconnectError)
	This probably means the server terminated abnormally
	before or while processing the request.
	from gems/sequel-5.10.0/lib/sequel/adapters/postgres.rb:152:in `block in execute_query'
	from gems/sequel-5.10.0/lib/sequel/database/logging.rb:38:in `log_connection_yield'
	from gems/sequel-5.10.0/lib/sequel/adapters/postgres.rb:152:in `execute_query'
	from gems/sequel-5.10.0/lib/sequel/adapters/postgres.rb:140:in `block in execute'
	from gems/sequel-5.10.0/lib/sequel/adapters/postgres.rb:116:in `check_disconnect_errors'
	from gems/sequel-5.10.0/lib/sequel/adapters/postgres.rb:140:in `execute'
	from gems/sequel-5.10.0/lib/sequel/adapters/postgres.rb:487:in `_execute'
	from gems/sequel-5.10.0/lib/sequel/adapters/postgres.rb:305:in `block (2 levels) in execute'
	from gems/sequel-5.10.0/lib/sequel/adapters/postgres.rb:509:in `check_database_errors'
	from gems/sequel-5.10.0/lib/sequel/adapters/postgres.rb:305:in `block in execute'
	from gems/sequel-5.10.0/lib/sequel/database/connecting.rb:253:in `block in synchronize'
	from gems/sequel-5.10.0/lib/sequel/connection_pool/threaded.rb:91:in `hold'
	from gems/sequel-5.10.0/lib/sequel/database/connecting.rb:253:in `synchronize'
	from gems/sequel-5.10.0/lib/sequel/adapters/postgres.rb:305:in `execute'
	from gems/sequel-5.10.0/lib/sequel/dataset/actions.rb:1083:in `execute'
	from gems/sequel-5.10.0/lib/sequel/adapters/postgres.rb:601:in `fetch_rows'
	from gems/sequel-5.10.0/lib/sequel/dataset/actions.rb:151:in `each'
	from gems/sequel-5.10.0/lib/sequel/dataset/actions.rb:442:in `map'
	from gems/sequel-5.10.0/lib/sequel/dataset/actions.rb:442:in `map'
	from gems/sequel-5.10.0/lib/sequel/adapters/shared/postgres.rb:1182:in `pg_class_relname'
	from gems/sequel-5.10.0/lib/sequel/adapters/shared/postgres.rb:733:in `tables'
	from seq.rb:7:in `<main>'
zsh: exit 1     ruby seq.rb
```

My guess is ActiveRecord doesn't try to close connections it doesn't own in the forked process.

**Comment by eregon** [comment]
I wonder though, what closes the connection in the forked process with Sequel?
A finalizer from `pg`? An `at_exit` hook?
Using `exit!` in the forked process seems to let the script succeed with Sequel.

**Comment by jeremyevans** [comment]
I believe the ActiveRecord change is related to: https://github.com/rails/rails/pull/31173/commits/f32cff5563f2188e657aa2fd9f8513f0da4a49ca

Let's just say `@connection.socket_io.reopen(IO::NULL)` is not exactly what I consider supported.

I believe the GC finalizer on the PG::Connection does a protocol level disconnect over the socket before closing it.  From my understanding, that's fairly common in ruby database drivers.

In any case, yes, I think if you are using `fork`, you should know to use `exec` or `exit!` if you want to avoid the issue.
